README.md
# estree-walkie

[![Latest version](https://img.shields.io/npm/v/estree-walkie)
 ![Dependency status](https://img.shields.io/librariesio/release/npm/estree-walkie)
](https://www.npmjs.com/package/estree-walkie)
[![Coverage](https://codecov.io/gh/prantlf/estree-walkie/branch/master/graph/badge.svg)](https://codecov.io/gh/prantlf/estree-walkie)

Walks a JavaScript [AST] compatible with [ESTree].

* [ESM], [CJS] and [UMD] modules offering an [API] for programmatic usage in [Node.js], [Deno] and web browser.
* Ignores keys in AST nodes pointing to parent nodes. (Often needed and would cause a stack overflow.)
* Includes TypeScript types.
* No dependencies.
* Tiny size - 926 B minified, 435 B gzipped, 390 B brotlied.
* [Blazingly fast].

## Synopsis

```js
import { parse } from 'meriyah'
import { walk } from 'estree-walkie'

const program = parse('...', { module: true, next: true })
const identifiers = []

walk(program, {
  Identifier(node) {
    identifiers.push(node.name)
  }
})
```

## API

Exported methods:

    walk(node, visitor, fallback, state)
    walkAtOffset(node, offset, visitor, fallback, state)
    walkAtPosition(node, line, column, visitor, fallback, state)
    walkInRange(node, start, end, visitor, fallback, state)
    walkBetweenLines(node, first, last, visitor, fallback, state)

Arguments:

* `node` - [AST] node to start walking with
* `visitor` - object with visiting handlers named by node types
* `fallback` - handler for nodes not matched by the visitor (optional)
* `state` - custom value passed to each handler (optional)
* `offset` - source code character offset to walk the [AST] around
* `start` and `ens` - source code character range to walk the [AST] within
* `line` and `column` - source code position to walk the [AST] around
* `first` and `last` - source code line range to walk the [AST] within

Visiting handler:

    handler(node, state, parent) => boolean?
    { enter?: handler, exit?: handler }

* `node` - currently visited [AST] node
* `state` - custom value passed to the walking method (optional)
* `node` - parent of the currently visited [AST] node

If an object with `enter` and/or `exit` methods is used as a visiting handler, `enter` will be caled before visiting children and `exit` after visiting children of the current node.

If a handler retuns `false` children of the current node will not be visited.

Walking can be interrupted by throwing an error and ignoring the thrown type in error handling.

## Examples

Collect only identifiers used in the outer scope:

```js
import { parse } from 'meriyah'
import { walk } from 'estree-walkie'

const program = parse('...', { module: true, next: true })
const identifiers = []

const avoid = () => false
walk(program, {
  FunctionDeclaration: avoid,
  ClassDeclaration: avoid,
  CallExpression: avoid,
  FunctionExpression: avoid,
  ArrowFunctionExpression: avoid,
  Identifier(node) {
    identifiers.push(node.name)
  }
})
```

Collect variables declared at all scopes and assign `node.parent` to all nodes:

```js
import { parse } from 'meriyah'
import { walk } from 'estree-walkie'

const program = parse('...', { module: true, next: true, loc: true })
const variables = []

const assignParent = (node, parent) => node.parent = parent
walk(astRoot, line, column, {
  VariableDeclarator(node, state, parent) {
    if (node.type !== 'Identifier') {
      throw new Error(`unsupported declarator at ${node.loc.start.line}:${node.loc.start.column + 1}`)
    }
    state.push(node.name)
    assignParent(node, parent)
  }
}, (node, _state, parent) => assignParent(node, parent), variables)
```

Check if there is an identifier at a specific position and remember its parent:

```js
import { parse } from 'meriyah'
import { walk } from 'estree-walkie'

const program = parse('...', { module: true, next: true, loc: true })
let identified

try {
  walkAtLocation(astRoot, line, column, {
    Identifier(node, _state, parent) {
      identified = parent
      throw 0
    }
  });
} catch (err) {
  if (typeof err !== 'number') throw err
}
```

## Node.js usage

Make sure that you use [Node.js] >= 8.3. Install the `estree-walkie` package locally using your favourite package manager:

```
npm i estree-walkie
pnpm i estree-walkie
yarn add estree-walkie
```

Either import methods from the [API] using the [CJS] module:

```js
const { walk, walkAtOffset, walkAtPosition, walkInRange, walkBetweenLines } = require('estree-walkie')
...
```

Or import methods from the [API] using the [ESM] module:

```js
import { walk, walkAtOffset, walkAtPosition, walkInRange, walkBetweenLines } from 'estree-walkie'
...
```

## Browser usage

You can either compile this module to your application bundle, or load it directly to the browser. For the former, you would install this package and import methods from the [API] in the same way as for the [Node.js usage]. For the latter, you would either refer to the module installed locally, or to the module from the [UNPKG CDN], for example:

    node_modules/estree-walkie/dist/index.umd.min.js
    https://unpkg.com/estree-walkie@0.0.1/dist/index.min.mjs

The following modules are available in the `dist` directory:

| Name               | Type                       |
| ------------------ | -------------------------- |
| `index.cjs`        | [CJS] module, not minified |
| `index.mjs`        | [ESM] module, not minified |
| `index.min.mjs`    | [ESM] module, minified     |
| `index.umd.js`     | [UMD] module, not minified |
| `index.umd.min.js` | [UMD] module, minified     |

Either import methods from the [API] using the [ESM] module:

```html
<script type=module>
  import { walk, walkAtOffset, walkAtPosition, walkInRange, walkBetweenLines }
    from 'https://unpkg.com/estree-walkie@0.0.1/dist/index.min.mjs'
  ...
</script>
```

Or import methods from the [API] using the [UMD] module, which will set a global object `estree-walkie`:

```html
<script src=https://unpkg.com/estree-walkie@0.0.1/dist/index.umd.min.js></script>
<script>
  (() => {
    const { walk, walkAtOffset, walkAtPosition, walkInRange, walkBetweenLines } = estreeWalkie
    ...
  })()
</script>
```

If an [AMD] module loader is detected, the [UMD] module will return exports es expected:

```html
<script>
  require(['https://unpkg.com/estree-walkie@0.0.1/dist/index.umd.min.js'],
    ({ walk, walkAtOffset, walkAtPosition, walkInRange, walkBetweenLines }) => {
      ...
    })
</script>
```

## Performance

Comparison with other libraries, which allow visiting [AST] nodes by their types:

    $ cd bench && pnpm i && node walk

    Validation:
      ✔ @babel/traverse (46,955 identifiers)
      ✔ estree-walker (46,955 identifiers)
      ✔ estree-walk (46,955 identifiers)
      ✔ estree-visitor (46,955 identifiers)
      ✘ acorn-walk (26,662 identifiers)
      ✔ ast-types (46,955 identifiers)
      ✔ astray (46,955 identifiers)
      ✔ estree-walkie (46,955 identifiers)

    Benchmark:
      @babel/traverse x 8.18 ops/sec ±7.00% (25 runs sampled)
      estree-walker   x 95.59 ops/sec ±2.63% (69 runs sampled)
      estree-walk     x 46.53 ops/sec ±1.57% (60 runs sampled)
      estree-visitor  x 61.86 ops/sec ±1.87% (64 runs sampled)
      acorn-walk      x 78.81 ops/sec ±0.92% (68 runs sampled)
      ast-types       x 3.65 ops/sec ±8.24% (14 runs sampled)
      astray          x 98.37 ops/sec ±0.93% (72 runs sampled)
      estree-walkie   x 131 ops/sec ±0.99% (76 runs sampled)

## Contributing

In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint and test your code using `npm test`.

## License

Copyright (c) 2022 Ferdinand Prantl

Licensed under the MIT license.

[AST]: https://astexplorer.net/
[ESTree]: https://github.com/estree/estree#readme
[Node.js]: https://nodejs.org/
[Deno]: https://deno.land/
[Blazingly fast]: #performance
[API]: #api
[Node.js usage]: #nodejs-usage
[UNPKG CDN]: https://unpkg.com/
[CJS]: https://blog.risingstack.com/node-js-at-scale-module-system-commonjs-require/#commonjstotherescue
[UMD]: https://github.com/umdjs/umd#readme
[AMD]: https://github.com/amdjs/amdjs-api/wiki/AMD
[ESM]: https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/#content-head


CHANGELOG.md
## [0.0.4](https://github.com/prantlf/estree-walkie/compare/v0.0.3...v0.0.4) (2022-07-08)


### Bug Fixes

* Ignore keys pointing to parent node, fix walkAtPosition ([1debacf](https://github.com/prantlf/estree-walkie/commit/1debacf5eeed99d42a8b376bb2998514876fb53f))

## [0.0.3](https://github.com/prantlf/estree-walkie/compare/v0.0.2...v0.0.3) (2022-07-06)


### Bug Fixes

* Fix the name of the CJS export ([5c2160c](https://github.com/prantlf/estree-walkie/commit/5c2160c880166d6ea296d59d86098381d8f6a6f4))

## [0.0.2](https://github.com/prantlf/estree-walkie/compare/v0.0.1...v0.0.2) (2022-07-06)


### Bug Fixes

* Fix typo in an exported function in TypeScript types ([f6287c2](https://github.com/prantlf/estree-walkie/commit/f6287c2f986ef956967f21f2dcbe67b44dde58dd))

## 0.0.1 (2022-07-06)

Initial release.


test/skip.js
import { deepStrictEqual, strictEqual } from 'assert'
import tehanu from 'tehanu'
import { walk } from 'estree-walkie'
import { parse } from 'meriyah'

const parseTest = source => parse(source, { module: true, next: true })
const test = tehanu(import.meta.url)

test('should skip children if parent dictates', () => {
  const program = parseTest(`
    export function hello(props) {
      if (props.disabled) {
        var name = props.name || 'world'
        var verb = props.informal ? 'Yo' : 'Hello'
        return verb + ' , ' + name
      }
    }
  `)

  let count = 0
  walk(program, {
    IfStatement() {
      return false
    },
    Identifier(_node) {
      count++
    }
  })

  // would be 12 without skip
  strictEqual(count, 2)
})

test('should skip Node children of terminal', () => {
  const program = parseTest(`
    let foo = 'bar'
    if (bar) baz = 'foo'
  `)

  let count = 0
  const idents = []

  walk(program, {
    Identifier(node) {
      count++
      idents.push(node.name)
    },
    VariableDeclaration(_node) {
      return false
    }
  })

  strictEqual(count, 2)
  deepStrictEqual(idents, ['bar', 'baz'])
})

test('should skip children of block', () => {
  const program = parseTest(`
    let foo = 'bar'
    if (bar) baz = 'foo'
  `)

  let count = 0
  const idents = []

  walk(program, {
    Identifier(node) {
      count++
      idents.push(node.name)
    },
    IfStatement(_node) {
      return false
    }
  })

  strictEqual(count, 1)
  deepStrictEqual(idents, ['foo'])
})

test('should skip children on enter', () => {
  const program = parseTest(`
    let foo = 'bar'
    if (bar) baz = 'foo'
  `)

  let count = 0
  const idents = []

  walk(program, {
    Identifier(node) {
      count++
      idents.push(node.name)
    },
    IfStatement: {
      enter(_node) {
        return false
      }
    }
  })

  strictEqual(count, 1)
  deepStrictEqual(idents, ['foo'])
})

test('should skip children on exit', () => {
  const program = parseTest(`
    let foo = 'bar'
    if (bar) baz = 'foo'
  `)

  let count = 0
  const idents = []

  walk(program, {
    Identifier(node) {
      count++
      idents.push(node.name)
    },
    IfStatement: {
      exit(_node) {
        return false
      }
    }
  })

  // skip() on exit is too late
  strictEqual(count, 3)
  deepStrictEqual(idents, ['foo', 'bar', 'baz'])
})


test/state.js
import { deepStrictEqual, strictEqual } from 'assert'
import tehanu from 'tehanu'
import { walk } from 'estree-walkie'
import { parse } from 'meriyah'

const parseTest = source => parse(source, { module: true, next: true })
const test = tehanu(import.meta.url)

test('maintain state across visitor levels', () => {
  const program = parseTest(`
    export function hello(props) {
      var name = props.name || 'world'
      return 'Howdy, ' + name
    }
  `)

  const seen = []
  walk(program, {
    FunctionDeclaration(_node, state) {
      state.parent = 'hello'
      seen.push('function')
    },
    VariableDeclaration(_node, state) {
      strictEqual(state.parent, 'hello')
      state.variable = 'name'
      seen.push('variable')
    },
    ReturnStatement(_node, state) {
      strictEqual(state.parent, 'hello')
      strictEqual(state.variable, 'name')
      seen.push('return')
    }
  }, undefined, {})

  deepStrictEqual(seen, ['function', 'variable', 'return'])
})

test('accept initial state value', () => {
  const state = { count: 0 }
  const program = parseTest(`
    export function hello(props) {
      var name = props.name || 'world'
      return 'Howdy, ' + name
    }
  `)

  walk(program, {
    Identifier(_n, s) {
      strictEqual(s, state)
      s.count++
    }
  }, undefined, state)

  strictEqual(state.count, 6)
})


test/default.js
import { deepStrictEqual, ok, strictEqual } from 'assert'
import tehanu from 'tehanu'
import { walk } from 'estree-walkie'
import { parse } from 'meriyah'

function parseTest(source) {
  return parse(source, { module: true, next: true })
}

const test = tehanu(import.meta.url)

test('should be able to visit base node', () => {
  let seen = false
  const program = parseTest(`var hello = 'world'`)
  walk(program, {
    Program(node) {
      ok(node === program)
      seen = true
    }
  })
  ok(seen)
})

test('should traverse child nodes recursively', () => {
  const seen = []
  const program = parseTest(`
    const INFORMAL = true
    export function testing(props) {
      let name = props.name || props.fullname
      return INFORMAL ? whaddup(name) : greet(name)
    }
  `)

  walk(program, {
    Identifier(node) {
      seen.push(node.name)
    }
  })

  deepStrictEqual(seen, [
    'INFORMAL',
    'testing', 'props',
    'name', 'props', 'name', 'props', 'fullname',
    'INFORMAL', 'whaddup', 'name', 'greet', 'name'
  ])
})

test('should support visitors with enter/exit methods', () => {
  const state = { count: 0 }
  const program = parseTest(`
    function hello(props) {
      return props.name || 'world'
    }
  `)

  walk(program, {
    Program: {
      enter(_n, s) {
        strictEqual(s, state)
        strictEqual(s.count, 0)
      },
      exit(_n, s) {
        strictEqual(s, state)
        strictEqual(s.count, 4)
      }
    },
    Identifier(_n, s) {
      strictEqual(s, state)
      s.count++
    }
  }, undefined, state)

  strictEqual(state.count, 4)
})

test('should pass parent node to visitors', () => {
  const state = { count: 0 }
  const program = parseTest(`
    function hello(props) {
      return props.name || 'world'
    }
  `)

  walk(program, {
    Program(_n, s, p) {
      strictEqual(p, undefined)
      strictEqual(s, state)
      strictEqual(s.count, 0)
      s.count++
    },
    Identifier(_n, s, p) {
      strictEqual(typeof p, 'object')
      strictEqual(s, state)
      s.count++
    }
  }, {}, state)

  strictEqual(state.count, 5)
})


src/index.js
function walkInner(node, test, visitor, fallback, state, parent) {
  if (test && !test(node)) return

  const visit = visitor[node.type] || fallback
  if (visit) {
    if (typeof visit === 'function') {
      if (visit(node, state, parent) === false) return false
    } else if (visit.enter) {
      if (visit.enter(node, state, parent) === false) return false
    }
  }

  for (const key in node) {
    const child = node[key]
    if (child && typeof child === 'object') {
      if (Array.isArray(child)) {
        for (const item of child) {
          if (item && typeof item === 'object') {
            walkInner(item, test, visitor, fallback, state, node)
          }
        }
      } else if (child.type && child !== parent) {
        walkInner(child, test, visitor, fallback, state, node)
      }
    }
  }

  if (visit && visit.exit) return visit.exit(node, state, parent)
}

export function walk(node, visitor, fallback, state) {
  return walkInner(node, undefined, visitor, fallback, state)
}

export function walkAtOffset(node, offset, visitor, fallback, state) {
  const test = node => {
    const { range } = node
    return range[0] <= offset && range[1] > offset
  }
  return walkInner(node, test, visitor, fallback, state)
}

export function walkInRange(node, start, end, visitor, fallback, state) {
  const test = node => {
    const { range } = node
    return range[0] < end && range[1] > start
  }
  return walkInner(node, test, visitor, fallback, state)
}

export function walkAtPosition(node, line, column, visitor, fallback, state) {
  const test = node => {
    const { loc } = node
    const { start, end } = loc
    return start.line <= line && end.line >= line &&
      (start.line !== line || end.line !== line ||
       start.column <= column && end.column > column)
  }
  return walkInner(node, test, visitor, fallback, state)
}

export function walkBetweenLines(node, first, last, visitor, fallback, state) {
  const test = node => {
    const { loc } = node
    return loc.start.line <= last && loc.end.line >= first
  }
  return walkInner(node, test, visitor, fallback, state)
}


src/index.d.ts
import * as AST from 'estree'

type Handler = (node: AST.Node, state: any, parent: AST.Node) => boolean | void

type Handlers = {
  enter?: Handler
  exit?: Handler
}

type Visitor = {
  [key in (keyof AST)]?: Handler | Handlers
}

export function walk(node: AST.Node, visitor: Visitor, fallback?: Handler | Handlers, state?: any): void
export function walkAtOffset(node: AST.Node, offset: number, visitor: Visitor, fallback?: Handler | Handlers, state?: any): void
export function walkAtPosition(node: AST.Node, line: number, column: number, visitor: Visitor, fallback?: Handler | Handlers, state?: any): void
export function walkInRange(node: AST.Node, start: number, end: number, visitor: Visitor, fallback?: Handler | Handlers, state?: any): void
export function walkBetweenLines(node: AST.Node, first: number, last: number, visitor: Visitor, fallback?: Handler | Handlers, state?: any): void


bench/walk.js
import { readFileSync } from 'fs'
import { strictEqual } from 'assert'
import benchmark from 'benchmark'
import babel from '@babel/traverse'
import { walk as estree } from 'estree-walker'
import eswalk from 'estree-walk'
import esvisit from 'estree-visitor'
import { simple as acorn } from 'acorn-walk'
import { visit as asttypes } from 'ast-types'
import { walk as astray } from 'astray'
import { walk as eswalkie } from 'estree-walkie'

const { default: traverse } = babel
const { Suite } = benchmark

const fixtures = {
  '@babel/traverse': 'fixtures/babel.json',
  'estree-walker': 'fixtures/estree.json',
  'estree-walk': 'fixtures/estree.json',
  'estree-visitor': 'fixtures/estree.json',
  'acorn-walk': 'fixtures/acorn.json',
  'ast-types': 'fixtures/estree.json',
  'astray': 'fixtures/estree.json',
  'estree-walkie': 'fixtures/estree.json'
}

const walkers = {
  '@babel/traverse': tree => {
    let count = 0
    traverse(tree, {
      noScope: true,
      Identifier() {
        ++count
      }
    })
    return count
  },
  'estree-walker': tree => {
    let count = 0
    estree(tree, {
      enter(n) {
        if (n.type === 'Identifier') {
          ++count
        }
      }
    })
    return count
  },
  'estree-walk': tree => {
    let count = 0
    eswalk(tree, {
      Identifier() {
        ++count
      }
    })
    return count
  },
  'estree-visitor': tree => {
    let count = 0
    esvisit(tree, {
      enter: function(n) {
        if (n.type === 'Identifier') {
          ++count
        }
      }
    })
    return count
  },
  'acorn-walk': tree => {
    let count = 0
    acorn(tree, {
      Identifier() {
        ++count
      }
    })
    return count
  },
  'ast-types': tree => {
    let count = 0
    asttypes(tree, {
      visitIdentifier(path) {
        ++count
        this.traverse(path)
      }
    })
    return count
  },
  'astray': tree => {
    let count = 0
    astray(tree, {
      Identifier() {
        ++count
      }
    })
    return count
  },
  'estree-walkie': tree => {
    let count = 0
    eswalkie(tree, {
      Identifier() {
        ++count
      }
    })
    return count
  }
}

const inputs = {}

console.log('\nValidation: ')
Object.keys(walkers).forEach(name => {
  const input = JSON.parse(readFileSync(fixtures[name], 'utf8'))
  const idents = walkers[name](input)

  inputs[name] = input
  try {
    strictEqual(idents, 46955, 'saw 46,955 identifiers')
    console.log('  ✔', `${name} (46,955 identifiers)`)
  } catch {
    console.log('  ✘', `${name} (${idents.toLocaleString()} identifiers)`)
  }
})

console.log('\nBenchmark:')
const suite = new Suite().on('cycle', ({ target }) => console.log('  ' + target))

Object.keys(walkers).forEach(name => {
  const walker = walkers[name]
  const input = inputs[name]
  suite.add(name + ' '.repeat(15 - name.length), () => walker(input))
})

suite.run()
